include/rocksdb/utilities/backupable_db.h

```diff
@@ -280,24 +280,43 @@ struct BackupFileInfo {
 typedef uint32_t BackupID;
 
 struct BackupInfo {
-  BackupID backup_id;
+  BackupID backup_id = 0U;
   // Creation time, according to GetCurrentTime
-  int64_t timestamp;
+  int64_t timestamp = 0;
 
   // Total size in bytes (based on file payloads, not including filesystem
   // overheads or backup meta file)
-  uint64_t size;
+  uint64_t size = 0U;
 
   // Number of backed up files, some of which might be shared with other
   // backups. Does not include backup meta file.
-  uint32_t number_files;
+  uint32_t number_files = 0U;
 
   // Backup API user metadata
   std::string app_metadata;
 
-  // Backup file details, if requested
+  // Backup file details, if requested with include_file_details=true
   std::vector<BackupFileInfo> file_details;
 
+  // DB "name" (a directory in the backup_env) for opening this backup as a
+  // read-only DB. This should also be used as the DBOptions::wal_dir, such
+  // as by default setting wal_dir="". See also env_for_open.
+  // This field is only set if include_file_details=true
+  std::string name_for_open;
+
+  // An Env(+FileSystem) for opening this backup as a read-only DB, with
+  // DB::OpenForReadOnly or similar. This field is only set if
+  // include_file_details=true. (The FileSystem in this Env takes care
+  // of making shared backup files openable from the `name_for_open` DB
+  // directory.) See also name_for_open.
+  //
+  // This Env might or might not be shared with other backups. To work
+  // around DBOptions::env being a raw pointer, this is a shared_ptr so
+  // that keeping either this BackupInfo, the BackupEngine, or a copy of
+  // this shared_ptr alive is sufficient to keep the Env alive for use by
+  // a read-only DB.
+  std::shared_ptr<Env> env_for_open;
```

Comment:
This seems very dangerous to me.  If env_for_open is set to Env::Default (for example), then won't the default env be deleted once this is out of scope?

Cn this be a FileSystem instead of an Env?  Are there operations in the Env that you require that are not in the FS?

Reply:
env_for_open is a field, so for an API user to reset it to some raw ptr they would have to use .reset() on a raw ptr they don't control the lifetime of. What's the new hazard here? Also this is an "output" struct for our API. Why would an API user be writing to these fields at all?

> Cn this be a FileSystem instead of an Env?

Then the caller has unnecessary complexity of creating their own temporary Env. What's the benefit?